cpp/include/cudf/io/arrow_io_source.hpp

```diff
@@ -0,0 +1,137 @@
+/*
+ * Copyright (c) 2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include "datasource.hpp"
+
+#include <arrow/buffer.h>
+#include <arrow/filesystem/filesystem.h>
+#include <arrow/io/file.h>
+#include <arrow/io/memory.h>
+
+#include <memory>
+#include <string>
+
+namespace cudf::io {
+/**
+ * @addtogroup io_datasources
+ * @{
+ * @file
+ */
+
+/**
+ * @brief Implementation class for reading from an Apache Arrow file. The file
+ * could be a memory-mapped file or other implementation supported by Arrow.
+ */
+class arrow_io_source : public datasource {
+  /**
+   * @brief Implementation for an owning buffer where `arrow::Buffer` holds the data.
+   */
+  class arrow_io_buffer : public buffer {
+    std::shared_ptr<arrow::Buffer> arrow_buffer;
+
+   public:
+    explicit arrow_io_buffer(std::shared_ptr<arrow::Buffer> arrow_buffer)
+      : arrow_buffer(arrow_buffer)
+    {
+    }
+    [[nodiscard]] size_t size() const override { return arrow_buffer->size(); }
+    [[nodiscard]] uint8_t const* data() const override { return arrow_buffer->data(); }
+  };
+
+ public:
+  /**
+   * @brief Constructs an object from an Apache Arrow Filesystem URI
+   *
+   * @param arrow_uri Apache Arrow Filesystem URI
+   */
+  explicit arrow_io_source(std::string const& arrow_uri)
```

Comment:
Same to all other member functions. Maybe considering separating class declaration from definition

Reply:
I want to keep this class separate from libcudf, but we don't have to take that route.
@davidwendt thoughts on keeping the implementation in libcudf?

Reply:
I'm not sure I understand what you mean by keeping this separate from libcudf.
I think Yunsong means moves the implementation to `cpp/src/io/arrow_io_source.cpp` and keep only the declarations in this header file. I agree with that approach.

Reply:
I meant that I will need to compile cpp/src/io/arrow_io_source.cpp as part of libcudf. Not an issue, just different from what I imagined in the beginning.